### PR TITLE
Update the bootstrap script to install wget-ssl

### PR DESCRIPTION
**Pull request by matteodelabre**
Entware recently renamed its `wget` package to `wget-ssl`, and made `wget` a virtual package provided both by `wget-ssl` and `wget-nossl` (see <https://github.com/openwrt/packages/pull/13613>). The latter seems to be favored by the package manager, so doing `opkg install wget` will actually install the version of wget that does not support TLS.

This PR changes the bootstrap script to ensure that the TLS-enabled version of wget gets installed. To test it, run the new bootstrap script on a clean system and make sure that no message saying “TLS certificate validation not implemented” is ever printed out.

### CODE REVIEW

**Comment by Eeems** [comment]
image

**Comment by Eeems** [comment]
image
Looks like it doesn't like running when it's already installed. Probably pre-existing. I had to remove the extra line from `opkg.conf` after it ran.

EDIT: Actually it looks like our logic for determining what branch we are on isn't working properly right now.

**Comment by Eeems** [comment]
```bash
 branch="$(sed -E 's|^src/gz\s+toltec\s+https://toltec-dev.org/([[:alnum:]]*).*$|\1|;t;d' "$opkg_conf_path" | head -n1)"
```
Likely should be something like this:
```bash
 branch="$(sed -E 's|^src/gz\s+toltec-rmall\s+https://toltec-dev.org/([[:alnum:]]*).*$|\1|;t;d' "$opkg_conf_path" | head -n1)"
```

**Comment by matteodelabre** [comment]
@Eeems Good catch, thanks! I updated the script with this change.

**Comment by Eeems** [comment]
It seems to be adding a duplicate entry now for the device specific entry. I did confirm that there was only one line previously.
image

File: scripts/bootstrap/bootstrap
Comment (Eeems) [inline scripts/bootstrap/bootstrap]: Perhaps get-branch should allow checking for specific items (i.e. `rm1`, `rm2`, and `rmall`)? And where it's used should be a little smarter about which one it's checking and replacing?

**Review by Eeems — DISMISSED** [review]
```bash
    # Make sure that Opkg is configured to fetch the chosen Toltec branch
    sed -i '/^src\/gz\s\+toltec[a-z-]*\s\+/d' "$opkg_conf_path"
    echo "src/gz toltec-rmall https://toltec-dev.org/$toltec_branch/rmall" >> "$opkg_conf_path"
    echo "src/gz toltec-$device_arch https://toltec-dev.org/$toltec_branch/$device_arch" >> "$opkg_conf_path"
```
This looks to be the culprit.

**Comment by matteodelabre** [comment]
Thanks for investigating this. In #356 I changed the bootstrap script to prevent it from running over an existing install, what do you think of that approach? Some pros and cons I can see:

* (+) Makes the script simpler (we don’t have to worry about each specific state the system could be in)
* (+) Eliminates some bugs (the issues you’re describing here + issue #332)
* (+) Having a script called “bootstrap” do other things than installing Toltec could be confusing to users.
* (-) We loose the ability to switch branches by using `bootstrap`—but I think this would be better handled as a standalone script, maybe building `opkg.conf` from a set of config files as you mentioned previously, with a lesser risk for bugs
* (-) We also loose the ability to migrate existing installs to new configs by using this script. But migration can always be handled separately, for example in an upgrade script of the `toltec-bootstrap` package. Also, it is always possible to migrate by uninstalling/reinstalling Toltec (although this might not always be desirable).

**Comment by Eeems** [comment]
I like the `toltecctl` idea. I do think that will take longer for us to test and merge though, so we might still want to resolve this and merge it.

**Comment by matteodelabre** [comment]
> ```shell
>     # Make sure that Opkg is configured to fetch the chosen Toltec branch
>     sed -i '/^src\/gz\s\+toltec[a-z-]*\s\+/d' "$opkg_conf_path"
>     echo "src/gz toltec-rmall https://toltec-dev.org/$toltec_branch/rmall" >> "$opkg_conf_path"
>     echo "src/gz toltec-$device_arch https://toltec-dev.org/$toltec_branch/$device_arch" >> "$opkg_conf_path"
> ```
> 
> This looks to be the culprit.

Ah, yes, the sed expression does not match source lines with digits in their name. I’ll fix that.

File: scripts/bootstrap/bootstrap
Comment (matteodelabre) [reply scripts/bootstrap/bootstrap]: It’s a fair point, but installs using different branches for rm1, rm2 and rmall should be fairly rare, if they exist at all. I think the current behavior is fine for now, we’ll be able to implement more complex behaviors when we move this logic to a specific script for handling branches.

**Comment by LinusCDE** [comment]
Bump

**Comment by matteodelabre** [comment]
Since #356 was merged first, this is now a fairly lightweight PR. To test it, instead of re-running the bootstrap script over an existing install, it is now necessary to uninstall using `toltecctl` then reinstall with the new boostrap.

**Review by LinusCDE — DISMISSED** [review]
Works like a charm. Tested it on my rM 1 and it god indeed rid of the "TLS not implemented" warning. Depending on whether #372 or this repo are merged first, one needs a version bump.

Just tag me in discord once one is bumped so I can re-approve the change.

<details>

<summary>Log (followed the previous testplan)</summary>

```shell
reMarkable: /home/root/ rm bs
reMarkable: /home/root/ nano bs
reMarkable: /home/root/ chmod +x bs
reMarkable: /home/root/ toltec_branch=testing bash bs
INFO:  Fetching secure wget
INFO:  Creating /home/root/.entware and mounting to /opt
INFO:  Installing Entware to /opt
2021-06-22 01:01:38 URL:https://bin.entware.net/armv7sf-k3.2/installer/opkg [752476/752476] -> "/opt/bin/opkg" [1]
2021-06-22 01:01:38 URL:https://bin.entware.net/armv7sf-k3.2/installer/opkg.conf [171/171] -> "/opt/etc/opkg.conf" [1]
2021-06-22 01:01:39 URL:https://bin.entware.net/armv7sf-k3.2/installer/ld-2.27.so [134428/134428] -> "/opt/lib/ld-2.27.so" [1]
2021-06-22 01:01:41 URL:https://bin.entware.net/armv7sf-k3.2/installer/libc-2.27.so [1247408/1247408] -> "/opt/lib/libc-2.27.so" [1]
2021-06-22 01:01:41 URL:https://bin.entware.net/armv7sf-k3.2/installer/libgcc_s.so.1 [50848/50848] -> "/opt/lib/libgcc_s.so.1" [1]
2021-06-22 01:01:42 URL:https://bin.entware.net/armv7sf-k3.2/installer/libpthread-2.27.so [92656/92656] -> "/opt/lib/libpthread-2.27.so" [1]
Downloading https://bin.entware.net/armv7sf-k3.2/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/entware
Installing entware-opt (227000-3) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/entware-opt_227000-3_all.ipk
Installing libgcc (8.4.0-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libgcc_8.4.0-11_armv7-3.2.ipk
Installing libc (2.27-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libc_2.27-11_armv7-3.2.ipk
Installing libssp (8.4.0-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libssp_8.4.0-11_armv7-3.2.ipk
Installing libpthread (2.27-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libpthread_2.27-11_armv7-3.2.ipk
Installing librt (2.27-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/librt_2.27-11_armv7-3.2.ipk
Installing libstdcpp (8.4.0-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libstdcpp_8.4.0-11_armv7-3.2.ipk
Installing entware-release (1.0-2) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/entware-release_1.0-2_all.ipk
Installing zoneinfo-asia (2021a-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/zoneinfo-asia_2021a-1_armv7-3.2.ipk
Installing zoneinfo-europe (2021a-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/zoneinfo-europe_2021a-1_armv7-3.2.ipk
Installing findutils (4.8.0-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/findutils_4.8.0-1_armv7-3.2.ipk
Installing terminfo (6.2-2) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/terminfo_6.2-2_armv7-3.2.ipk
Installing libpcre (8.44-4) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libpcre_8.44-4_armv7-3.2.ipk
Installing grep (3.6-1a) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/grep_3.6-1a_armv7-3.2.ipk
Installing locales (2.27-9) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/locales_2.27-9_armv7-3.2.ipk
Installing opkg (2021-03-15-5936c4f9-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/opkg_2021-03-15-5936c4f9-1_armv7-3.2.ipk
Installing entware-upgrade (1.0-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/entware-upgrade_1.0-1_all.ipk
Configuring libgcc.
Configuring libc.
Configuring libssp.
Configuring libpthread.
Configuring librt.
Configuring terminfo.
Configuring libpcre.
Configuring grep.
Configuring locales.
Entware uses separate locale-archive file independent from main system
Creating locale archive /opt/usr/lib/locale/locale-archive
Adding en_EN.UTF-8
Adding ru_RU.UTF-8
You can download locale sources from http://bin.entware.net/other/i18n_glib227.tar.gz
You can add new locales to Entware using /opt/bin/localedef.new
Configuring entware-upgrade.
Upgrade operations are not required.
Configuring opkg.
Configuring zoneinfo-europe.
Configuring zoneinfo-asia.
Configuring libstdcpp.
Configuring entware-release.
Configuring findutils.
Configuring entware-opt.
INFO:  Configuring the Toltec testing repository
Downloading https://bin.entware.net/armv7sf-k3.2/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/entware
Downloading https://toltec-dev.org/testing/rmall/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/toltec-rmall
Downloading https://toltec-dev.org/testing/rm1/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/toltec-rm1
Installing ca-certificates (20210119-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/ca-certificates_20210119-1_all.ipk
Installing wget-ssl (1.21.1-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/wget-ssl_1.21.1-1_armv7-3.2.ipk
Installing zlib (1.2.11-3) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/zlib_1.2.11-3_armv7-3.2.ipk
Installing libopenssl (1.1.1k-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libopenssl_1.1.1k-1_armv7-3.2.ipk
Configuring ca-certificates.
Configuring zlib.
Configuring libopenssl.
Configuring wget-ssl.
Downloading https://bin.entware.net/armv7sf-k3.2/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/entware
Downloading https://toltec-dev.org/testing/rmall/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/toltec-rmall
Downloading https://toltec-dev.org/testing/rm1/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/toltec-rm1
Installing toltec-bootstrap (0.0.6-1) to root...
Downloading https://toltec-dev.org/testing/rmall/toltec-bootstrap_0.0.6-1_rmall.ipk
Configuring toltec-bootstrap.
Please restart your SSH session or run 'exec bash -l' to use Toltec
INFO:  Please restart your SSH session or run 'exec bash -l' to use Toltec
INFO:  After each system upgrade, run 'toltecctl enable' to re-enable Toltec
```

</details>

**Comment by matteodelabre** [comment]
Thanks for your review. I merged #372 first and bumped this PR’s version. Also @Eeems it seems that we can’t merge this PR as long as you are requesting changes, even if Linus approves it.

**Review by LinusCDE — APPROVED** [review]

**Comment by matteodelabre** [comment]
@Eeems Turns out I can simply dismiss the review, sorry for the ping! (And thanks Linus for the tip.)

**Pull request merged**
